### PR TITLE
HostFS: Retain File Handles in Save States

### PR DESCRIPTION
**Pull request by pips3**
### Description of Changes

[SAVEVERSION+]
Save host file handle and position information in save states, and reopen handles at the correct position when states are loaded.

Note: This is my first contribution, and I am still learning PCSX2's codebase, as well as best C++ practices. Criticism is appreciated!

### Rationale behind Changes

Currently, if a homebrew/game uses Host Filesystem to stream files, using save states can cause issues. For example if the streamed file is audio, it can cause audio to be desynced, stop entirely or even play the incorrect song. This is caused by open file handles no longer existing or the file offset pointer being different when loading a savestate.
### Suggested Testing Steps

@penguino118 has graciously created a homebrew elf that demonstrates this issue. The program streams audio from `GTA.wav`and `HMU.wav`. (Assets are copyright free/public domain)
[hostfs_handle_test.zip](https://github.com/user-attachments/files/17266999/hostfs_handle_test.zip)

Expected behaviour on master:

1. Boot `tutorial_06.elf`. Press X to begin.
2. Save a state while a song is playing.
3. Wait a few seconds for the song to progress, then load the state you just made.
4. The song will continue playing from the point it was at before the state was loaded. (Bad)
5. Press X to Switch song. Load the state again. The song you switched to will continue to play instead of the one the state was saved on. (Bad!)
6. Reset the elf, do NOT press X to begin. Load the state. The song will endlessly stutter. (Bad!!)

Expected behaviour on this branch:

1. Open `tutorial_06.elf`. Press X to begin.
2. Save a state while a song is playing.
3. Wait a few seconds for the song to progress, then load the state you just made.
4. The song will play from the point it was at when the state was made. (Good)
5. Press X to Switch song. Load the state again. It will play the song that the state was saved on. (Good)
6. Reset the elf, do NOT press X to begin. Load the state. The song will play. (Good)

### CODE REVIEW

File: pcsx2/IopBios.cpp [hunk 1]
```diff
@@ -1387,3 +1415,57 @@ namespace R3000A
...
+
+			R3000A::fileHandle handle;
+			Freeze(handle.flags);
+			FreezeString(handle.full_path);
+			Freeze(handle.mode);
+			R3000A::handles.push_back(handle);
+
+			//reopen the file
+			IOManFile* file = NULL;
+			R3000A::HostFile::open(&file, handle.full_path, handle.flags, handle.mode);
```
Comment (F0bes) [inline pcsx2/IopBios.cpp:1450]: Can you check the return of this open call or check if `file` is `nullptr`? It's probably best to remove the handle if this is the case. 
You can run into issues if you load a state after renaming / removing a file that was previously saved.

Also I'd add a warning in the console with the file path via `Console.Warning(...)` so people know why their state isn't working as intended.

**Review by F0bes — CHANGES_REQUESTED** [review]
Thanks, this is a great addition. I just noticed a couple issues and warnings :)

File: pcsx2/IopBios.cpp [hunk 2]
```diff
@@ -1387,3 +1415,57 @@ namespace R3000A
...
+	const int firstfd = R3000A::ioman::firstfd;
+	int handleCount = R3000A::handles.size();
+	Freeze(handleCount);
+
+	for (int i = 0; i < handleCount; i++)
+	{
+		if (IsLoading())
+		{
+			//load the parameters for opening the file
+			long pos;
```
[the remark this answers is not in the document]
Comment (pips3) [reply pcsx2/IopBios.cpp]: Ooo that's not good, thanks. I'll have to remember to test on linux as well in the future :P

File: pcsx2/IopBios.cpp [hunk 3]
```diff
@@ -1387,3 +1415,62 @@ namespace R3000A
...
+			}
+			R3000A::handles[i].fd_index = R3000A::ioman::allocfd(file) - firstfd;
+
+			//seek file to position when saved
+			file->lseek(pos, SEEK_SET);
+		}
+		else
+		{
+			//save the current file position
+			u32 fd = R3000A::handles[i].fd_index;
```
Comment (F0bes) [inline pcsx2/IopBios.cpp:1464]: can be const, otherwise looks good to me.

**Review by kamfretoz — APPROVED** [review]
LGTM, tested on both Master and PR.

**Review by JordanTheToaster — APPROVED** [review]

**Pull request merged**